library/std/src/gc.rs

```diff
@@ -175,11 +201,76 @@ pub fn thread_registered() -> bool {
     unsafe { bdwgc::GC_thread_is_registered() != 0 }
 }
 
+// Ring buffer of objects to be finalized.
+//
+// This is considered a root for the Gc because the 'to-be-finalised' objects are otherwise
+// unreachable and must be kept alive until their finaliser has run.
+static FIN_Q: Mutex<VecDeque<Gc<()>>> = Mutex::new(VecDeque::new());
+
+// Number of objects to grow the FIN_Q by each time it reaches capacity.
+pub const FIN_Q_BUFFER_SIZE: usize = 2048;
+
+// The function pointer passed down the queue to be executed on the finalization thread.
+//
+// This is not polymorphic because type information about the object inside the box is not known
+// during collection. However, it is enough to use the () type because we ensure that during
+// allocation that it points to the correct drop_in_place fn for the underlying value.
+type Finalizer = unsafe fn(*mut GcBox<()>);
+
+/// Called by the collector during the STW pause, where the collector holds the
+/// global allocator lock. This function therefore *must not* perform any heap allocation or the
+/// program would deadlock.
+unsafe extern "C" fn finalizer_stub(object: *mut u8, _meta: *mut u8) -> u32 {
+    let mut q = FIN_Q.lock().unwrap();
+    // The ring buffer must have enough space to push the object. If not, this would trigger a call
+    // to realloc which would deadlock.
+    assert!(q.len() != q.capacity());
+    q.push_back(Gc::from_inner(NonNull::new_unchecked(object as *mut GcBox<()>)));
+    drop(q);
+    return 1;
+}
+
+pub fn process_fin_q() {
+    loop {
+        let object = FIN_Q.lock().unwrap().pop_front();
```

Comment:
This is interesting: I *think* we only have a 1 finalisation thread, right? But that thread still has to acquire a lock to do finalisation even though *we* know that no other thread can have gained that lock (outside of GC time).

If I've got that right, it suggests that we also want to *move* the "queue of queues" to the finalisation thread, so that it doesn't have to do any locking.

Reply:
> This is interesting: I think we only have a 1 finalisation thread, right? 

Correct

> But that thread still has to acquire a lock to do finalisation even though we know that no other thread can have gained that lock (outside of GC time).

But the `FIN_Q` is like a producer / consumer, so we need to lock because this could be popping from the queue at the same time that new allocations on the mutator push to the queue. Maybe I've misunderstood your suggestion.

Reply:
>But the FIN_Q is like a producer / consumer, so we need to lock because this could be popping from the queue at the same time that new allocations on the mutator push to the queue. Maybe I've misunderstood your suggestion.

The finaliser thread will be stopped during GC so at run-time I don't think it can ever actually race? But we will stay pay the lock/unlock costs even though we're not really doing anything that could ever race...

Reply:
> The finaliser thread will be stopped during GC so at run-time I don't think it can ever actually race? But we will stay pay the lock/unlock costs even though we're not really doing anything that could ever race...

Doh... You're right! This hadn't occurred to me. If I can then malloc / mmap this queue by side-stepping the Boehm allocator, I might be able to grow it inside the STW pause and remove locking altogether.